iiasa/message_ix
# Enhancing storage equations by considering `mode` and `lvl_temporal`

This PR addresses #632 by adding `mode` to the mapping set of storage technologies (`map_tec_storage`) and also to storage equations.
More details:
This change means re-initializing the existing set of `map_tec_storage` with new index sets. Based on the discussion in #631, this re-initializing happens automatically by modifications in the file `models.py`, only if the mapping set is empty. If the set is not empty, an exception is raised, telling the user that they need to update this set.
Moreover, this PR improves the documentation of storage section to reflect the above changes and more than that, adding required details for each equation.

## How to review

- create a branch from the submitted PR; load an existing scenario (from Westeros, national or global models); clone a new scenario without carrying the solution (using the argument `keep_solution=False`); solve the cloned scenario again. The cloned scenario should solve without an error, if there is no content in the set `map_tec_storage` in that scenario. Otherwise, there should be an error asking you to update this set.
- run  `test_feature_storage.py` from this PR on the main branch; it should fail.
- read the documentation of the storage section and migration notes, and make sure it's clear and complete.

## PR checklist

- [x] Test is enhanced (`test_feature_storage.py`) to reflect the new changes.
- [x] Expanded the documentation.
- [x] Update release and migration notes.
